### PR TITLE
Basic Datepicker Component

### PR DESCRIPTION
**Pull request by Zombiefruit**

### CODE REVIEW

File: demo/components-demo.tsx [hunk 1]
```diff
@@ -29,6 +28,9 @@ export class ComponentsDemo extends React.Component<{}, {}>{
                 <TreeViewDemo />
             </div>
             <hr />
+            <h1>{'<DatePicker />'}</h1>
```
Comment (tobich) [inline demo/components-demo.tsx]: Please adjust the demo section to the rest of the components. It's been changed meanwhile, methink.

File: demo/components/date-picker-demo.tsx [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import * as React from 'react';
+import { DatePicker } from '../../src';
+import {observable, action} from 'mobx';
+import {observer} from 'mobx-react';
+
+@observer
+export class DatePickerDemo extends React.Component<{}, {}> {
+    render () {
+        return <DatePicker placeholder="mm/dd/yyyy" />;
```
Comment (tobich) [inline demo/components/date-picker-demo.tsx]: DatePicker must be a controlled component, like the rest of the bunch. So it can't possibly work without some external state in the demo bound to the `value` and `onChange` props.

File: src/common/environment.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+export const isBrowser: boolean = typeof window !== 'undefined';
+
+// for server-side-rendering, we assume the client will support object-fit
+export const objectFitSupported: boolean = !isBrowser || ('object-fit' in document.createElement('img').style);
```
Comment (tobich) [inline src/common/environment.ts]: How is this related to the DatePicker?

File: src/components/date-picker/date-picker-dropdown.tsx [hunk 4]
```diff
@@ -0,0 +1,68 @@
+import * as React from 'react';
+import {getMonthNames, getMonthFromOffset} from './date-picker-helpers';
+import {observable, action, computed} from 'mobx';
+import {observer} from 'mobx-react';
+import {DatePickerGrid} from './date-picker-grid';
+import styles from './date-picker.st.css';
+
+export interface DatePickerDropdownProps {
+    date: Date;
```
Comment (tobich) [inline src/components/date-picker/date-picker-dropdown.tsx]: I suggest `value` for uniformity.

File: src/components/date-picker/date-picker-dropdown.tsx [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    goToNextMonth: React.EventHandler<React.SyntheticEvent<HTMLSpanElement>> = (event: React.SyntheticEvent<HTMLSpanElement>) => {
+        event.preventDefault();
+        const nextMonth: Date = getMonthFromOffset(new Date(this.date.getFullYear(), this.date.getMonth(), 1), 1);
+        this.setDateTo(nextMonth);
+    };
+
+    goToPrevMonth: React.EventHandler<React.SyntheticEvent<HTMLSpanElement>> = (event: React.SyntheticEvent<HTMLSpanElement>) => {
+        event.preventDefault();
+        event.stopPropagation();
```
Comment (tobich) [inline src/components/date-picker/date-picker-dropdown.tsx]: This doesn't appear in the previous method. They should be symmetrical. Also, why is preventDefault() not enough?

File: src/components/date-picker/date-picker-grid.tsx [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            dayArray.push(i);
+        }
+
+        return dayArray.map(day => <span className={[styles.calendarItem, styles.day].join(' ')} onMouseDown={this.onMouseDown} key={'DAY_' + day}
+                                              data-automation-id={'DAY_' + day}>{day}</span>);
+    }
+
+    @computed
+    get dayNames (): Array<JSX.Element> {
+        return getDayNames(this.props.startingDay).map((name, index) => <span className={[styles.calendarItem, styles.dayName].join(' ')} key={'DAY_NAME_' + index}
```
Comment (tyv) [inline src/components/date-picker/date-picker-grid.tsx]: day names also have to be prop, because of i18n or locale should be in props, and used as arg for `getDayNames `

File: src/components/date-picker/date-picker-grid.tsx [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        const eventTarget = event.target as HTMLElement;
+        this.props.onChange(eventTarget.textContent!)
+    };
+
+    render() {
+        return (
+                <div className={styles.calendar} data-automation-id="CALENDAR">
+                    {this.dayNames}
+                    {this.previousDays}
+                    {this.dayArray}
```
Comment (tyv) [inline src/components/date-picker/date-picker-grid.tsx]: probably it is `this.days` -- why to use `array` word in method name, and also it is multiple value.

File: src/components/date-picker/date-picker-dropdown.tsx [hunk 8]
```diff
@@ -0,0 +1,68 @@
+import * as React from 'react';
```
Comment (tyv) [inline src/components/date-picker/date-picker-dropdown.tsx]: I would love to see some universal `<Dropdown/>` implementation and then its usage in `<DatePickerDropdown/>`

File: src/components/date-picker/date-picker-helpers.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    // Days start from Sunday (Sunday = 0, Monday = 1, etc.)
+    const listOfDayNames: Array<string> = [];
+
+    for (let i = 0; i < 7; i++) {
+        let dayName = new Date(1970, 5, i + firstDay).toLocaleString(locale, { weekday: nameLength });
+        listOfDayNames.push(dayName);
+    }
+
+    return listOfDayNames;
+}
```
Comment (tyv) [inline src/components/date-picker/date-picker-helpers.ts]: `return Array.apply(null, {length: 7}).map(day => ...)`

File: src/components/date-picker/date-picker-helpers.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+export function getMonthNames (locale: string = 'en-US'): Array<string> {
+    const listOfMonthNames: Array<string> = [];
+    for (let i = 0; i < 12; i++) {
+        let monthName = new Date(1970, i, 1).toLocaleString(locale, { month: 'long' });
+        listOfMonthNames.push(monthName);
+    }
+    return listOfMonthNames;
```
Comment (tyv) [inline src/components/date-picker/date-picker-helpers.ts]: same as above

File: src/components/date-picker/date-picker.css [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+.headerArrowPrev {
+    transform: rotate(45deg);
+    margin-left: 24px;
+    border-left: arrowThickness solid color_MainTextBackground;
+}
+
+.calendar {
+    display: flex;
+    flex-wrap: wrap;
+    /*align-content: space-between;*/
```
Comment (tyv) [inline src/components/date-picker/date-picker.css]: forgot comment

File: src/components/date-picker/date-picker.css [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    height: calc(100% - headerHeight);
+    width: dropdownWidth;
+    padding: 30px;
+}
+
+.calendarItem {
+    flex: 0 0 14.2857%;
```
Comment (tyv) [inline src/components/date-picker/date-picker.css]: it is a very strange value

File: src/components/date-picker/date-picker-grid.tsx [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+@observer
+export class DatePickerGrid extends React.Component<DatePickerGridProps, {}> {
+    static defaultProps = {
+        startingDay: 0
+    };
+
+    @computed
+    get dayArray (): Array<JSX.Element> {
+        const dayArray: Array<number> = [];
+        for (let i = 1; i <= getDaysInMonth(this.props.date); i++) {
```
Comment (wtfil) [inline src/components/date-picker/date-picker-grid.tsx]: this will be callen for each loop iteration

File: src/components/date-picker/date-picker-grid.tsx [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+            dayArray.push(i);
+        }
+
+        return dayArray.map(day => <span className={[styles.calendarItem, styles.day].join(' ')} onMouseDown={this.onMouseDown} key={'DAY_' + day}
+                                              data-automation-id={'DAY_' + day}>{day}</span>);
```
Comment (wtfil) [inline src/components/date-picker/date-picker-grid.tsx]: @tyv there is a `fill` method for this
```new Array(getDaysInMonth(this.props.date)).fill().map(data => ....)```

Comment (wtfil) [inline src/components/date-picker/date-picker-grid.tsx]: also this find of formating is hard to read. All props should be either in one line or each property in separate line

File: src/components/date-picker/date-picker.tsx [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+                       onKeyDown={this.onKeyDown}
+                       onMouseDown={this.onMouseDown}
+                       onBlur={this.onBlur}
+                       onFocus={this.onFocus}
+                       onChange={this.onInputChange}
+                       value={this.inputValue}
+                       placeholder={this.props.placeholder}
+                       type="text"
+                       data-automation-id="DATE_PICKER_INPUT" />
+                {this.showDropdown ?
```
Comment (wtfil) [inline src/components/date-picker/date-picker.tsx]: formatting?

File: src/components/date-picker/index.ts [hunk 16]
```diff
@@ -0,0 +1,2 @@
+export * from './date-picker';
+export * from './date-picker-helpers';
```
Comment (wtfil) [inline src/components/date-picker/index.ts]: this should not be exported into component list

File: src/components/date-picker/date-picker-grid.tsx (hunk 14 above)
Comment (tyv) [inline src/components/date-picker/date-picker-grid.tsx]: @wtfil yes! I was trying to remember that, it is even nicer

File: src/components/date-picker/date-picker-dropdown.tsx (hunk 8 above)
Comment (tobich) [inline src/components/date-picker/date-picker-dropdown.tsx]: We will hvae our own Popup component soon

File: src/components/date-picker/date-picker-dropdown.tsx [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+import {observer} from 'mobx-react';
+import {DatePickerGrid} from './date-picker-grid';
+import styles from './date-picker.st.css';
+
+export interface DatePickerDropdownProps {
+    date: Date;
+    onChange (date: Date): void;
+}
+
+const monthNames = getMonthNames();
```
Comment (tobich) [inline src/components/date-picker/date-picker-dropdown.tsx]: i18n coming later (soon)

File: src/components/date-picker/date-picker-dropdown.tsx [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+
+@observer
```
Comment (tobich) [inline src/components/date-picker/date-picker-dropdown.tsx]: the date picker grid is actually a part which will be better optimized with mobx.

File: src/components/date-picker/date-picker.tsx [hunk 19]
```diff
@@ -0,0 +1,106 @@
+import * as React from 'react';
+import {DatePickerDropdown} from './date-picker-dropdown';
+import {observable, action} from 'mobx';
+import {observer} from 'mobx-react';
+import {SyntheticEvent} from "react";
+import {KeyCodes} from '../../common/key-codes';
```
Comment (tobich) [inline src/components/date-picker/date-picker.tsx]: @AviVahl

File: src/components/date-picker/date-picker.tsx [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+        return this.date;
+    }
+
+    @observable date: Date = this.props.date ? this.props.date : new Date();
+    @observable inputValue: string = this.props.date ? this.props.date.toDateString() : '';
+    @observable showDropdown: boolean = this.props.showDropdown ? this.props.showDropdown : false;
+
+    // Called with possibly invalid string from the input
+    @action updateStateFromString = (input: string): void => {
+        if (this.isDateValid(input) || input === '') {
```
Comment (tobich) [inline src/components/date-picker/date-picker.tsx]: `|| input === ''` should be either tested or removed

File: src/components/date-picker/date-picker-grid.tsx (hunk 6 above)
Comment (tobich) [inline src/components/date-picker/date-picker-grid.tsx]: We're doing i18n later

File: src/components/date-picker/date-picker.css (hunk 12 above)
Comment (tobich) [inline src/components/date-picker/date-picker.css]: Should be documented (1/7)

File: src/components/date-picker/date-picker-dropdown.tsx [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+        this.setDateTo(previousMonth);
+    };
+
+    render() {
+        return (
+            <div>
+                <div className={styles.dropdownArrowWrapper}><div className={styles.dropdownArrow} /></div>
+                <div className={styles.dropdown} data-automation-id="DATE_PICKER_DROPDOWN">
+                    <div className={styles.header}>
+                        <span className={[styles.headerArrow, styles.headerArrowPrev].join(' ')} onMouseDown={this.goToPrevMonth} data-automation-id="PREV_MONTH_BUTTON"></span>
```
Comment (Zombiefruit) [inline src/components/date-picker/date-picker-dropdown.tsx]: Exactly, onClick fires after the blur happens

Comment (tyv) [inline src/components/date-picker/date-picker-dropdown.tsx]: it is ugly. 
you have to keep focus while clicks are inside dropdown, and dropdown shouldn't be closed until you're clicking inside it. (and yes, buttons are better than spans)

File: src/common/date-helpers.ts [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+    const listOfDayNames: Array<string> = [];
+
+    // For each of the seven days of the week
+    for (let i = 0; i < 7; i++) {
+        let dayName = new Date(1970, 5, i + firstDay).toLocaleString(locale, { weekday: nameLength });
+        listOfDayNames.push(dayName);
+    }
+
+    return listOfDayNames;
+}
```
Comment (tyv) [inline src/common/date-helpers.ts]:
```ts
export function getDayNames (firstDay: number = 0, nameLength: string = 'short', locale: string = 'en-US'): Array<string> {
  return new Array(7)
    .fill()
    .map((_, index) => 
      new Date(1970, 5, firstDay + index).toLocaleString('en-US', { weekday: 'short' })
    )
}
```

File: src/common/date-helpers.ts [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+    const listOfMonthNames: Array<string> = [];
+
+    // For each of the twelve months of the year
+    for (let i = 0; i < 12; i++) {
+        let monthName = new Date(1970, i, 1).toLocaleString(locale, { month: 'long' });
+        listOfMonthNames.push(monthName);
+    }
+    return listOfMonthNames;
+}
+
```
Comment (tyv) [inline src/common/date-helpers.ts]: same as above

File: src/components/date-picker/date-picker.tsx [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+    value: Date;
+    onChange (value: Date): void;
+    isDropdownVisible?: boolean;
+    placeholder?: string;
+}
+
+@observer
+export class DatePicker extends React.Component<DatePickerProps, {}>{
+    @observable inputValue: string = this.props.value ? this.props.value.toDateString() : '';
+    @observable isDropdownVisible: boolean = this.props.isDropdownVisible ? this.props.isDropdownVisible : false;
```
Comment (tyv) [inline src/components/date-picker/date-picker.tsx]: it could be ` = Boolean(this.props.isDropdownVisible)`

File: src/components/date-picker/date-picker.tsx [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+const invalidDate: string = 'Invalid Date';
+
+export interface DatePickerProps {
+    value: Date;
+    onChange (value: Date): void;
+    isDropdownVisible?: boolean;
+    placeholder?: string;
+}
+
+@observer
```
Comment (tyv) [inline src/components/date-picker/date-picker.tsx]: mobx here is used only for simple flag/value storing, do we really need it here?

File: src/components/date-picker/day.tsx [hunk 26]
```diff
@@ -0,0 +1,49 @@
+import * as React from 'react';
+import {computed} from 'mobx';
+import {observer} from 'mobx-react';
+
+const styles = require('./date-picker.st.css').default;
+
+export interface DayProps {
+    day: number | string;
+    selected?: boolean;
+    currentDay?: boolean;
```
Comment (tyv) [inline src/components/date-picker/day.tsx]: maybe `isCurrent`, `isSelected`?

File: src/components/date-picker/day.tsx [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+    focused?: boolean;
+    partOfPrevMonth?: boolean;
+    partOfNextMonth?: boolean;
```
Comment (tyv) [inline src/components/date-picker/day.tsx]: `fromPreviousMonth`, `fromNextMonth` would be better naming I hope

File: src/components/date-picker/day.tsx [hunk 28]
```diff
@@ -0,0 +1,49 @@
...
+    dataAutomationId?: string;
+
```
Comment (tyv) [inline src/components/date-picker/day.tsx]: why space?

File: src/components/date-picker/day.tsx [hunk 29]
```diff
@@ -0,0 +1,49 @@
...
+        } else if (this.props.selected) {
+            return [styles.calendarItem, styles.day, styles.selectedDay].join(' ');
+        } else if (this.props.currentDay) {
+            return [styles.calendarItem, styles.day, styles.currentDay].join(' ');
+        } else if (this.props.partOfNextMonth || this.props.partOfPrevMonth) {
+            return [styles.calendarItem, styles.greyDay].join(' ');
+        } else {
+            return [styles.calendarItem, styles.day].join(' ');
+        }
+    }
```
Comment (tyv) [inline src/components/date-picker/day.tsx]: 1. is fact that day could be focused and current or other combinations is covered? 
2. grayDay - is bad name. It should be 'otherMonthDay' or smth like that
3. I think each modifier prop should be represented as separate classname, even if 2 classnames will have same styling, that will help working with this component in projects.
4. does it really have to be mobx computed value?

File: src/components/date-picker/day.tsx [hunk 30]
```diff
@@ -0,0 +1,49 @@
...
+
+    onMouseDown: React.EventHandler<React.SyntheticEvent<HTMLSpanElement>> = (event: React.SyntheticEvent<HTMLSpanElement>) => {
```
Comment (tyv) [inline src/components/date-picker/day.tsx]: it is still span, don't we have `HTMLElement` which is more common

File: src/components/date-picker/dropdown.tsx [hunk 31]
```diff
@@ -0,0 +1,131 @@
...
+
+export interface DropdownProps {
+    value: Date;
+    onChange(date: Date): void;
+}
+
+const monthNames = getMonthNames();
+
+@observer
+export class Dropdown extends React.Component<DropdownProps, {}>{
```
Comment (tyv) [inline src/components/date-picker/dropdown.tsx]: I would separate everything related to calendar grid to `<Grid/>` component, so when we will have `<Dropdown/>` or `<Popup />` component we can use `<Grid/>` here and it will be less painful refactoring.

File: src/components/date-picker/dropdown.tsx [hunk 32]
```diff
@@ -0,0 +1,131 @@
...
+    @action
+    setDate (date: Date) {
+        this.date = date;
+    }
+
+    @action
+    setDay = (day: number) => {
+        const date = new Date(this.date.getFullYear(), this.date.getMonth(), day);
+        this.props.onChange(date);
+    };
```
Comment (tyv) [inline src/components/date-picker/dropdown.tsx]: `setDay` is bad name because it has side effect. 
I would use `onChange` which calls `this.props.onChange` and `setDate`

File: src/components/date-picker/dropdown.tsx [hunk 33]
```diff
@@ -0,0 +1,131 @@
...
+
+        return dayArray.map((day: number) => {
+            return (
+                <Day day={day}
+                     onSelect={this.setDay}
+                     dataAutomationId={'DAY_' + day}
+                     key={'DAY_' + day} />
+            );
+        });
+    }
```
Comment (tyv) [inline src/components/date-picker/dropdown.tsx]:
```tsx
return new Array(getDaysInMonth(this.date)).fill().map((_, index: number) => {
     const id = `DAY_${index}`
     return (
        <Day 
            day={index}
            onSelect={this.setDay}
            dataAutomationId={id}
            key={id} />         
     );
 })
```

File: src/components/date-picker/dropdown.tsx [hunk 34]
```diff
@@ -0,0 +1,131 @@
...
+
+    @computed
+    get dayNames (): Array<JSX.Element> {
+        return getDayNames().map((name: string, index: number) => {
+            return (
+                <span className={[styles.calendarItem, styles.dayName].join(' ')}
+                      key={'DAY_NAME_' + index}
+                      data-automation-id={'DAY_NAME_' + name.toUpperCase()}>
```
Comment (tyv) [inline src/components/date-picker/dropdown.tsx]: you can use same id as key and automation

File: src/components/date-picker/dropdown.tsx [hunk 35]
```diff
@@ -0,0 +1,131 @@
...
+                     key={'DAY_' + day} />
+            );
+        });
+    }
+
+    @computed
+    get dayNames (): Array<JSX.Element> {
+        return getDayNames().map((name: string, index: number) => {
+            return (
+                <span className={[styles.calendarItem, styles.dayName].join(' ')}
```
Comment (tyv) [inline src/components/date-picker/dropdown.tsx]: I believe joins looks very imperative

```tsx
const {calendarItem, dayName} = styles;
...
className={`${calendarItem} ${dayName}`}
```

File: src/components/date-picker/dropdown.tsx [hunk 36]
```diff
@@ -0,0 +1,131 @@
...
+const styles = require('./date-picker.st.css').default;
+
+export interface DropdownProps {
+    value: Date;
+    onChange(date: Date): void;
+}
+
+const monthNames = getMonthNames();
+
+@observer
```
Comment (tyv) [inline src/components/date-picker/dropdown.tsx]: I still don't see problem with rendering calendar grid and if it has any performance issues. 
no any issues since jquery.

Comment (tyv) [inline src/components/date-picker/dropdown.tsx]: @tobich ^^

File: src/components/date-picker/dropdown.tsx [hunk 37]
```diff
@@ -0,0 +1,131 @@
...
+        });
+    }
+
+    @computed
+    get previousDays (): Array<JSX.Element> {
+        const previousDays: Array<JSX.Element> = [];
+        const lastDayOfPrevMonth: number = getDaysInMonth(getMonthFromOffset(this.date, -1));
+        const numberOfDaysToDisplay: number = getNumOfPreviousDays(this.date);
+
+        for (let i = (lastDayOfPrevMonth - numberOfDaysToDisplay) + 1; i <= lastDayOfPrevMonth; i++) {
```
Comment (tyv) [inline src/components/date-picker/dropdown.tsx]: this will calculate each iteration without reason (both are consts)
this can be easily transformed to `map`, see examples in comments above

File: src/components/date-picker/day.tsx (hunk 28 above)
Comment (tobich) [inline src/components/date-picker/day.tsx]: what space?

File: src/components/date-picker/day.tsx (hunk 29 above)
Comment (tobich) [inline src/components/date-picker/day.tsx]: Agree with the point (2)

File: src/components/date-picker/dropdown.tsx (hunk 35 above)
Comment (tobich) [inline src/components/date-picker/dropdown.tsx]: Imperative? what?

File: src/components/date-picker/dropdown.tsx (hunk 37 above)
Comment (tobich) [inline src/components/date-picker/dropdown.tsx]: I agree with the first comment.
The map is not necessarily better here.

File: src/components/date-picker/dropdown.tsx (hunk 32 above)
Comment (tobich) [inline src/components/date-picker/dropdown.tsx]: onChange is indeed better name. It doesn't have to be an action, btw.

File: src/components/date-picker/dropdown.tsx (hunk 33 above)
Comment (tobich) [inline src/components/date-picker/dropdown.tsx]: I suggest another approach: you can push the JSX elements directly into the dayArray and return it at the end.

File: src/components/date-picker/dropdown.tsx (hunk 36 above)
Comment (tobich) [inline src/components/date-picker/dropdown.tsx]: We decided to do this kind of optimizations over all repeaters.

File: src/components/date-picker/day.tsx (hunk 27 above)
Comment (tobich) [inline src/components/date-picker/day.tsx]: partOf* is pretty clear to me

File: src/components/date-picker/day.tsx (hunk 26 above)
Comment (tobich) [inline src/components/date-picker/day.tsx]: current is ok

File: src/components/date-picker/date-picker.tsx (hunk 24 above)
Comment (tobich) [inline src/components/date-picker/date-picker.tsx]: No need for type conversion, actually.
This should be inside constructor and within `@autorun` call. The same in the previous line.

File: src/components/date-picker/day.tsx (hunk 29 above)
Comment (wtfil) [inline src/components/date-picker/day.tsx]: agree with not having common classes. There is not difficulty to use `,` in css

File: test/components/date-picker.spec.tsx [hunk 38]
```diff
@@ -0,0 +1,430 @@
...
+    const clientRenderer = new ClientRenderer();
+
+    afterEach(() => clientRenderer.cleanup());
+
+    const JANUARY_FIRST = new Date(2017, 0, 1);
+    const FEBRUARY_FIRST = new Date(2017, 1, 1);
+    const MARCH_FIRST  = new Date(2017, 2, 1);
+    const DECEMBER_FIRST = new Date(2017, 11, 1);
+
+    const simulateChange = (domElememt: Element, value: string) => {
```
Comment (tobich) [inline test/components/date-picker.spec.tsx]: Please use test-drive-react/trigger.change() instead

File: test/components/date-picker.spec.tsx [hunk 39]
```diff
@@ -0,0 +1,430 @@
...
+            const {select, waitForDom} = clientRenderer.render(<DatePickerDemo />);
+
+            const datePickerInput = select(datePickerInputId);
+            simulateChange(datePickerInput!, '2017-02-01');
+            simulate.keyDown(datePickerInput, { keyCode: KeyCodes.ENTER });
+
+            await waitForDom(() => expect(datePickerInput).to.have.value('Wed Feb 01 2017'));
+        });
+
+        it('writes into the date picker input field, focuses elsewhere, and expects the date picker input to have the proper value', async () => {
```
Comment (tobich) [inline test/components/date-picker.spec.tsx]: This is just another version of the previous test. It should be unit-test

File: test/components/date-picker.spec.tsx [hunk 40]
```diff
@@ -0,0 +1,430 @@
...
+
+    describe('A Typical User', () => {
+        it('writes into the date picker input field, presses enter, and expects the date picker input to have the proper value', async () => {
+            const {select, waitForDom} = clientRenderer.render(<DatePickerDemo />);
+
+            const datePickerInput = select(datePickerInputId);
+            simulateChange(datePickerInput!, '2017-02-01');
+            simulate.keyDown(datePickerInput, { keyCode: KeyCodes.ENTER });
+
+            await waitForDom(() => expect(datePickerInput).to.have.value('Wed Feb 01 2017'));
```
Comment (tobich) [inline test/components/date-picker.spec.tsx]: Consider using some external value indicator in the demo

File: test/components/date-picker.spec.tsx [hunk 41]
```diff
@@ -0,0 +1,430 @@
...
+    });
+
+    describe('The Dropdown', () => {
+        const dayNames: Array<string> = getDayNames();
+        const days: Array<string> = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28'];
+
+        it('should display the days for a fixed month', async () => {
+            const {select, waitForDom} = clientRenderer.render(<DatePickerDemo isDropdownVisible={true} value={FEBRUARY_FIRST} />);
+
+            await waitForDom(() => dayNames.forEach(name => expect(select(datePickerDropdownId)).to.contain.text(name)));
```
Comment (tobich) [inline test/components/date-picker.spec.tsx]: No need for two separate waitForDom I guess

File: src/components/date-picker/day.tsx (hunk 29 above)
Comment (wtfil) [inline src/components/date-picker/day.tsx]: btw, you can use classNames as is
`"calendarItem day"` instead of `[styles.calendarItem, styles.day].join(' ')`

File: src/components/date-picker/date-picker.tsx (hunk 25 above)
Comment (Zombiefruit) [inline src/components/date-picker/date-picker.tsx]: Currently removing mobx from the top-level datepicker

File: src/common/date-helpers.ts (hunk 22 above)
Comment (tyv) [inline src/common/date-helpers.ts]: it is much harder to read with `for` loop, I don't see any reason making it harder to read.
implementation is ok, I agree, readability is bad, extra variable out of loop is bad, etc..

File: src/components/date-picker/day.tsx (hunk 28 above)
Comment (tyv) [inline src/components/date-picker/day.tsx]: empty space

File: src/components/date-picker/dropdown.tsx (hunk 34 above)
Comment (Zombiefruit) [inline src/components/date-picker/dropdown.tsx]: In cases where the days of week are displayed as single letters (e.g. S M T W T F S) this will result in two identical keys.

File: src/components/date-picker/day.tsx (hunk 29 above)
Comment (tyv) [inline src/components/date-picker/day.tsx]: > Agree with the point (2)

@tobich how about other 3?

File: src/components/date-picker/dropdown.tsx (hunk 36 above)
Comment (tyv) [inline src/components/date-picker/dropdown.tsx]: I thought we decided that we use mobx only when we can have noticable profit

File: src/components/date-picker/dropdown.tsx (hunk 33 above)
Comment (tyv) [inline src/components/date-picker/dropdown.tsx]: why don't you want to use map and chaining?

File: src/components/date-picker/dropdown.tsx (hunk 35 above)
Comment (tyv) [inline src/components/date-picker/dropdown.tsx]: template string is more declarative way to divide 2 string variables with space then `join()`.
anyway, as @wtfil said we can use just `className = "calendarItem dayName"`

File: src/components/date-picker/dropdown.tsx (hunk 37 above)
Comment (tyv) [inline src/components/date-picker/dropdown.tsx]: which get us back to few comments above

File: test/components/date-picker.spec.tsx (hunk 41 above)
Comment (Zombiefruit) [inline test/components/date-picker.spec.tsx]: Yeah, not sure why I did it that way...

File: src/common/date-helpers.ts (hunk 22 above)
Comment (Zombiefruit) [inline src/common/date-helpers.ts]: I think this is quite subjective. Personally, I find the for-loops more readable. I did change some of the variable names so perhaps it's a bit easier to read now.

File: src/components/date-picker/day.tsx (hunk 28 above)
Comment (Zombiefruit) [inline src/components/date-picker/day.tsx]: Carry-over from core3 days. Removed

File: src/common/date-helpers.ts (hunk 22 above)
Comment (tyv) [inline src/common/date-helpers.ts]: it is subjective, yes: or you using imperative style, or functional. 
`.map` and chaining are abstractions made to avoid loops. It is fact. Use this abstraction or not is subjective. 
I think if language gives you abstraction you have to it.

File: demo/components-demo.tsx (hunk 1 above)
Comment (tobich) [inline demo/components-demo.tsx]: Still different, @Zombiefruit !

File: src/common/date-helpers.ts (hunk 22 above)
Comment (tobich) [inline src/common/date-helpers.ts]: I believe creating array just for the sake of index count is making the code more obscure and less efficient with no actual gain. If we were using lodash, i would perhaps use `lodash.times().map()`. But the current approach is totally fine.

File: src/components/date-picker/dropdown.tsx (hunk 36 above)
Comment (tobich) [inline src/components/date-picker/dropdown.tsx]: By "we" I meant @nadavov :-)

File: src/components/date-picker/dropdown.tsx (hunk 35 above)
Comment (tobich) [inline src/components/date-picker/dropdown.tsx]: Imperative means it has side-effects. join() is as declarative as template strings.

File: src/components/date-picker/day.tsx (hunk 29 above)
Comment (tyv) [inline src/components/date-picker/day.tsx]: why this thread is ignored?

File: src/components/date-picker/day.tsx [hunk 42]
```diff
@@ -0,0 +1,48 @@
...
+            return [styles.calendarItem, styles.day, styles.selectedDay].join(' ');
+        } else if (this.props.selected) {
+            return [styles.calendarItem, styles.day, styles.selectedDay].join(' ');
+        } else if (this.props.currentDay) {
+            return [styles.calendarItem, styles.day, styles.currentDay].join(' ');
+        } else if (this.props.partOfNextMonth || this.props.partOfPrevMonth) {
+            return [styles.calendarItem].join(' ');
+        } else {
+            return [styles.calendarItem, styles.day].join(' ');
+        }
```
Comment (tyv) [inline src/components/date-picker/day.tsx]: can those be just strings? It should work if you'll use SBComponent
